### PR TITLE
webhook_features: Request review from stale reviewers

### LINKED ISSUE #11: Ping stale ACK reviewers?
Would it be reasonable to ping reviewers if they left an ACK that became stale after a push?

### PR DESCRIPTION
**Pull request by aureleoules**
Attempt to fix #11.

This PR will make DrahtBot request reviews from stale reviews after every new push (those who left an ACK on a previous commit).

~~Currently in draft because this PR depends on https://github.com/XAMPPRocky/octocrab/pull/277.~~

### CODE REVIEW

File: conflicts/Cargo.lock [hunk 1]
```diff
@@ -771,7 +818,7 @@ dependencies = [
 [[package]]
 name = "octocrab"
 version = "0.17.0"
-source = "git+https://github.com/XAMPPRocky/octocrab?branch=master#9a677bf39167477e0b9dfa298dae03b3d6c19160"
+source = "git+https://github.com/XAMPPRocky/octocrab?branch=master#cc93dd39f0085154be4750863f8197d7637f644e"
```
Comment (MarcoFalke) [inline conflicts/Cargo.lock:821]: no need to bump here?

File: webhook_features/src/features/summary_comment.rs [hunk 2]
```diff
@@ -70,7 +70,7 @@ impl Feature for SummaryCommentFeature {
                 let pr_number = payload["number"]
                     .as_u64()
                     .ok_or(DrahtBotError::KeyNotFound)?;
-                refresh_summary_comment(ctx, repo, pr_number).await?
+                refresh_summary_comment(ctx, repo, pr_number, action == "synchronize").await?
```
Comment (MarcoFalke) [inline webhook_features/src/features/summary_comment.rs:73]: So if there are two pushes in 10 minutes, there will be several requests?

File: conflicts/Cargo.lock (hunk 1 above)
Comment (aureleoules) [reply conflicts/Cargo.lock:821]: I just updated everything, is this an issue?
The conflicts crate uses the `util` crate so wouldn't it create conflicting crates?

File: webhook_features/src/features/summary_comment.rs (hunk 2 above)
Comment (aureleoules) [reply webhook_features/src/features/summary_comment.rs:73]: Yes, but it won't do anything the second time. You can only request a review from someone if they haven't left one yet or if they've left a review after a review request (from what I've observed).

File: conflicts/Cargo.lock (hunk 1 above)
Comment (MarcoFalke) [reply conflicts/Cargo.lock:821]: Every crate compiles all dependencies from its own lockfile for itself. So different lockfiles for different crates shouldn't be an issue, as long as the dependencies inside the crate don't conflict with each other.

Comment (aureleoules) [reply conflicts/Cargo.lock]: Fixed, thanks.

**Comment by aureleoules** [comment]
Rebased

File: webhook_features/Cargo.lock [hunk 3]
```diff
@@ -1143,7 +1199,7 @@ dependencies = [
 [[package]]
 name = "octocrab"
 version = "0.17.0"
-source = "git+https://github.com/XAMPPRocky/octocrab?branch=master#06c0152089a3d5ff8174885c2c7ff899815aab22"
+source = "git+https://github.com/XAMPPRocky/octocrab?branch=master#cc93dd39f0085154be4750863f8197d7637f644e"
```
Comment (MarcoFalke) [inline webhook_features/Cargo.lock:1202]: I had issues compiling the other bumps last night, so maybe just bump this one and leave the others as-is for now?

See 60733f041ac3bca95613755f0109613a6b90dd20

Comment (MarcoFalke) [reply webhook_features/Cargo.lock:1202]: Done in f7591062ae0dfc8713ac63b8fa4af235c3f944a7. Feel free to rebase and drop the lock changes completely.

Comment (aureleoules) [reply webhook_features/Cargo.lock]: Done

**Comment by MarcoFalke** [comment]
Personally I like the delayed ping, but I'll wait for others to chime in here

**Comment by aureleoules** [comment]
> Personally I like the delayed ping, but I'll wait for others to chime in here

For how long would you delay the ping? I don't really mind delaying it but by how long? 10-15mins or days?

**Comment by MarcoFalke** [comment]
"a few days" was suggested in https://github.com/MarcoFalke/DrahtBot/issues/11#issuecomment-1326815040, so maybe a week?

**Comment by aureleoules** [comment]
A delay of days/week would require DrahtBot to have a stored state in case it crashes or just needs to be updated. I think it's a bit of a hassle but if others think it's worth it then okay.

**Comment by MarcoFalke** [comment]
It could also be done "async" (like the :-1: feature), on the next (unrelated) update.

Another alternative would be to store the state in the comment (maybe inside a html comment), and then run another process to fetch those comments and run the update. Similar to `inactive_rebase`.

**Comment by kouloumos** [comment]
I think delaying is more appropriate. What I've seen from my behavior, if I've done a review, and someone later comes in with a thoughtful review, I'll wait a bit after the author has address those comments to give a chance to that reviewer to give his approval. Of course I can't wait for ever, but if that original reviewer has more comments on what has already been discussed/fixed, this might spare me one review cycle.

That said, my experience in the project is limited, and I would really like to hear from more experience contributors on how they are approaching this kind of stuff.

Apart from that, I have another idea (inspired by https://github.com/bitcoin/bitcoin/issues/26556) that might be a good match for this. It's a bit more complicate and it also makes the workflow more depended on GitHub, which I don't like, but I believe that it can be easily decoupled in the future.

The idea is using DrahtBot and Projects to display a list of all the stale reviews of each contributor. I haven't spend more than 10m looking into the implementation details of this but it seems doable. Whoever wants this functionality needs to create a Project and then [give access](https://docs.github.com/en/issues/planning-and-tracking-with-projects/managing-your-project/managing-access-to-your-projects) to DrahtBot. So whenever there is a stale review, DrahtBot could go and add a record to the table. By also including a date, this could be used as reference for the delay.

**Comment by MarcoFalke** [comment]
This was also suggested here https://github.com/MarcoFalke/DrahtBot/issues/11#issuecomment-1324013430, so maybe start a new thread with this idea?

**Comment by kouloumos** [comment]
> This was also suggested here [#11 (comment)](https://github.com/MarcoFalke/DrahtBot/issues/11#issuecomment-1324013430), so maybe start a new thread with this idea?

Thanks for pointing it out. I'll look a bit into the implementation details and do that.

**Comment by ajtowns** [comment]
> A delay of days/week would require DrahtBot to have a stored state in case it crashes or just needs to be updated.

The list of requested reviewers is already stored state? I was thinking something like:

```
if last_push_date > now - 3 days:
    for who in ackers:
         review_up_to_date = False
         if who.last_comment_date > last_push_date + 5 minutes: review_up_to_date = True
         if who.acked_commit_hash = last_push_commit_hash: review_up_to_date = True
         if who.id in requested_reviewers and review_up_to_date:
             unrequest_review(who)
         elif who.id not in requested_reviewers and not review_up_to_date:
             request_review(who)
```

**Comment by MarcoFalke** [comment]
I think the question was more about how to *trigger* the bot. If there is no webhook after 3 days, currently it won't be running.

Maybe there can be a slow-running background thread that cycles all pulls in a loop?

**Comment by MarcoFalke** [comment]
I think async-only is better than nothing. I might pick this up unless your are still working on this?

**Comment by aureleoules** [comment]
> I might pick this up unless your are still working on this?

Go ahead and pick this up.

**Comment by MarcoFalke** [comment]
thx, did something in 0a9955b46b0a7f9277628094df28fb12f2cdffb1

**Pull request closed without merging**